service/matching/matching_engine.go

```diff
@@ -615,6 +615,9 @@ pollLoop:
 		resp, err := e.recordWorkflowTaskStarted(ctx, requestClone, task)
 		if err != nil {
 			switch err.(type) {
+			case *serviceerror.Internal, *serviceerror.DataLoss,
+				*serialization.DeserializationError, *serialization.SerializationError:
```

Comment:
are `serialization.DeserializationError` and `serialization.SerializationError` translated across rpc boundaries? I think only serviceerrors are...

Reply:
Turns out they aren't. Thanks for pointing this out. Did some digging and verified that this is the case as seen inside a `UnaryHandler`: 

>  from interceptor.go
> // If a UnaryHandler returns an error, it should either be produced by the
> // status package, or be one of the context errors. Otherwise, gRPC will use
> // codes.Unknown as the status code and err.Error() as the status message of the
> // RPC.


Now, this begs the question of how we still want to catch non-service errors (such as errors close to the persistence layer) inside `matching`. Instead of checking for the exact error messages inside `matching` as returned by the rpc, I propose to conduct the check inside `history` and convert serialization based errors into service errors.

This shall require us to have serialization based errors as a new serviceerror. I think this is pretty neat since we would catch more of these errors properly. Thoughts?

cc - @dnr @yycptt

Reply:
Note that there are two "serviceerror" packages, the public one in api-go, and the internal one in common/serviceerror. The latter is imported as "serviceerror**s**" (the last "s" for "server").

I think the two options are:
1. Add a new serialization error type to common/serviceerror (the internal one) and use that. It would be a subtype of codes.DataLoss.
2. Just convert serialization errors to the existing (public) serviceerror.DataLoss.

(1) is more information but I'm not sure it adds anything over just serviceerror.DataLoss so I might lean to (2)